### ISSUE TITLE
How does Hierarchy work?

### ISSUE THREAD

**Issue by koichirose** [OWASP/rbac#54]
TL:DR; I'm not sure how the hierarchy works :)

I've been running some tests on rbac.
I have 3 roles (member, collaborator, admin).
member has basic permissions
collaborator should have member permissions plus some others (created with parent: member).
admin should have collaborator permissions plus some others (created with parent: collaborator).
The 'root' role has the 'root' permission, has no parents and no children.

Here's what I came up to (my debugging output only, plus resulting table):

```
truncate table permissions
truncate table rolepermissions
truncate table roles
INSERT INTO `phprbac_permissions` (`ID`, `Lft`, `Rght`, `Title`, `Description`) VALUES (1, 0, 1, "root", "root");
INSERT INTO `phprbac_rolepermissions` (`RoleID`, `PermissionID`, `AssignmentDate`) VALUES (1, 1, UNIX_TIMESTAMP());
INSERT INTO `phprbac_roles` (`ID`, `Lft`, `Rght`, `Title`, `Description`) VALUES (1, 0, 1, "root", "root");

created role 'member' with description 'Registered User' and parent ''. New id: '2'

created permission 'write_site_review' with description 'Write review for a site' and parent ''. New id: '2'
assigned permission '2' to role '2'

created role 'collaborator' with description 'Collaborator user' and parent '2'. New id: '3'

created permission 'access_admin' with description 'Access the admin panel' and parent ''. New id: '3'
assigned permission '3' to role '3'

created role 'admin' with description 'Admin user' and parent '3'. New id: '4'

created permission 'view_stats' with description 'View stats' and parent ''. New id: '4'
assigned permission '4' to role '4'

created permission 'manage_users' with description 'Manage Users' and parent ''. New id: '5'
assigned permission '5' to role '4'

created permission 'edit_users' with description 'Edit Users' and parent '5'. New id: '6'
assigned permission '6' to role '4'

created permission 'export_users' with description 'Export Users as CSV' and parent '5'. New id: '7'
assigned permission '7' to role '4'
```

tests:

```
INSERT INTO `phprbac_userroles` (`UserID`, `RoleID`, `AssignmentDate`) VALUES (123, 2, UNIX_TIMESTAMP()); /*member*/
INSERT INTO `phprbac_userroles` (`UserID`, `RoleID`, `AssignmentDate`) VALUES (456, 3, UNIX_TIMESTAMP()); /*collaborator*/
INSERT INTO `phprbac_userroles` (`UserID`, `RoleID`, `AssignmentDate`) VALUES (789, 4, UNIX_TIMESTAMP()); /*admin*/

member can access admin? true //should be false
collaborator can access admin? true //ok
admin can access admin? false //should be true
member can write site reviews? true //ok
collaborator can write site reviews? false //should be true
admin can write site reviews? false //should be true
```

Resulting tables:
permissions

```
ID Lft Rght Title Description
1   0   13  root    root
2   1   2   write_site_review   Write review for a site
3   3   4   access_admin    Access the admin panel
4   5   6   view_stats  View stats
5   7   12  manage_users    Manage Users
6   8   9   edit_users  Edit Users
7   10  11  export_users    Export Users as CSV
```

roles

```
ID Lft Rght Title Description
1   0   7   root    root
2   1   6   member  Registered User
3   2   5   collaborator    Collaborator user
4   3   4   admin   Admin user
```

rolepermissions

```
RoleID PermissionID AssignmentDate
1   1   1413214573
2   2   1413214573
3   3   1413214573
4   4   1413214573
4   5   1413214573
4   6   1413214573
4   7   1413214573
```

I also tried reversing the parents:
admin with no parent.
collaborator with admin parent.
member with collaborator parent.

result:

```
member can access admin? false //ok
collaborator can access admin? true //ok
admin can access admin? false //should be true
member can write site reviews? true //ok
collaborator can write site reviews? false //should be true
admin can write site reviews? false //should be true
```

What am I doing wrong?

**Comment by koichirose** [comment]
Nevermind, I think I solved it by keeping the root parent.
For future users with this issue:
root has no parent.
admin has root parent.
collaborator has admin parent.
member has collaborator parent.

This seems to fix all my issues :)

**Comment by abiusx** [comment]
This is just a sample hierarchy in place. You just need to have root as the base node.

**Comment by jburns131** [comment]
@koichirose: you are correct, the top level parent for both roles and permissions needs to be 'root' for PHP-RBAC to work as designed.

And this might help you understand the hierarchy structure: http://en.wikipedia.org/wiki/Nested_set_model
